### ISSUE TITLE
Decrypt CBC mode - The parameter is incorrect

### ISSUE THREAD

**Issue by junkbox1** [IRSgov/IDES-Data-Preparation-Dot-Net#23]
I am able to encrypt a packet, and then decrypt the response packet so long as ECB is used. When CBC is used, I am able to encrypt, but when I attempt to decrypt the response I get the error "The parameter is incorrect", thrown by rsa.Decrypt(payload, false); in AesManager.DecryptAesKey.

I am able to encrypt and decrypt a packet locally using CBC when I use my own public key during the encryption process, and since ECB works entirely I am not sure that it is a problem with my key-pair. 

Since the test environment only supports CBC, any help is urgently appreciated.

Thanks!

**Comment by DanielCrouch** [comment]
Try downloading the CBC testpacket1.zip file from https://www.irs.gov/businesses/corporations/ides-data-transmission-and-file-preparation. Inside that, there is a notification (840FOlY+TaVCSbgN8gl+v3Re8xcENaIX.zip) and notification_key.p12 (password is "password") file that you can use to test your CBC decryption. Does that work for you?

**Comment by junkbox1** [comment]
Yes, this works for me.

**Comment by DanielCrouch** [comment]
Ok, and how about if you encrypt a packet using the same notification_key.p12 for both the private and public keys and then try to decrypt the resulting package, again with the notification_key.p12?

**Comment by junkbox1** [comment]
That succeeded as well.

**Comment by DanielCrouch** [comment]
So it all seems to work fine with known keys and files. But if you submit a test packet to IDES, you aren't able to decrypt the notification, correct?

You may want to see if anything in this closed thread and see if anything applies to you. https://github.com/IRSgov/IDES-Data-Preparation-Dot-Net/issues/9

As mentioned in there, the only time I have been able to hit that incorrect parameter error was exporting a key in Windows 10. For some reason, the same source .jks key exported into a .p12 from Windows 10 just would not work for me. That was well before any of the CBC mode change came about, so it was all ECB. So it working with ECB but not with CBC is strange.

**Comment by junkbox1** [comment]
Yes that is exactly correct - something seems to go wrong only when I am decrypting a notification from the IRS in CBC mode, otherwise all works as expected. I know ECB works because I uploaded a test file to the production region and was able to decrypt the response that effectively said 'don't upload test files to the production region'.

I exported the public and private keys on a box running Server 2012 R2 Datacenter. I have also done all of the encryption/decryption on another box with the same OS.

Do you have some instructions for a preferred method by which I can export the keys? Maybe I can try that method on a machine with Win7 Enterprise instead of Server 2012.

**Comment by DanielCrouch** [comment]
No, there isn't a preferred method. Everyone has a different way since the CA may give out the files in a different format and the SSL gets installed in all kinds of different server and OS environments. The one I had tested with was a self signed certificate, it wasn't from a particular CA. Another thing you can try just as a test, is to see if you run into similar issues using OpenSSL to do the decryption. If you get an error with that method as well, it is a sure sign that some oddity exists with the key. I have only seen the Incorrect Parameter error reported a couple of times over the last couple of years so it isn't a common issue at all. I think that previous thread I mentioned was the first.

**Comment by junkbox1** [comment]
I read that the test environment is set up for a user account on the day the testing period starts, and that any changes made to that account after that day (eg password change, preferences) are NOT reflected in the test environment. Is this true for the key I uploaded as well? It has changed since the testing period has started.

**Comment by DanielCrouch** [comment]
Yes, that is correct. After the enrollment cutoff date, your account information is copied into the test environment. If you change your key after that point, that change only takes place in the production environment. There is no mechanism to move that to the test environment once the test window has started.

**Comment by junkbox1** [comment]
Alright, that explains it then.

So I am wondering, would I get the same responses in production as I would in test?  When I submitted a test file to production with an invalid element, the response I got denoted that it failed XSD validation, as expected. When I submit a valid file, all I get is 'don't submit test to production'. If I get the latter message, is it relatively safe to assume the file is OK, or is there some other processing that is not getting done on your end after your prod region reads that the DocTypeIndic is test? I hope that question makes sense.
